wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheConnection.java

```diff
@@ -257,10 +328,197 @@ private void returnConnectionBackToPool(StatefulRedisConnection <byte[], byte[]>
+    }
+  }
+
   // Used for unit testing only
   protected void setConnectionPools(GenericObjectPool<StatefulRedisConnection<byte[], byte[]>> readPool,
       GenericObjectPool<StatefulRedisConnection<byte[], byte[]>> writePool) {
     readConnectionPool = readPool;
     writeConnectionPool = writePool;
   }
+
+  private void setUpAuthenticationProperties(Properties properties) {
+    boolean iamAuthEnabled = CACHE_IAM_USERNAME.getString(properties) != null &&
+        !CACHE_IAM_USERNAME.getString(properties).trim().isEmpty();
+    if (iamAuthEnabled) {
+      setUpIamAuthentication(properties);
+    } else if (hasTraditionalCredentials(properties)) {
+      setupTraditionalAuthentication(properties);
+    }
+  }
+
+  private void setUpIamAuthentication(Properties properties) {
+    String hostname = extractHostnameFromCacheEndpoint(properties);
+    if (hostname != null && isElastiCacheEndpoint(hostname)) {
+      String username = CACHE_IAM_USERNAME.getString(properties);
+      if (username == null || username.isEmpty()) {
+        throw new RuntimeException("cacheIamUsername property is required for ElastiCache IAM authentication");
+      }
+      String region = CACHE_IAM_REGION.getString(properties);
+      if (region == null || region.isEmpty()) {
+        throw new RuntimeException("cacheIamRegion property is required for ElastiCache IAM authentication. " +
+            "Please set it in your connection properties (e.g., cacheIamRegion=us-west-2)");
+      }
+      properties.setProperty("cacheIamUsername", username);
+      properties.setProperty("cacheIamRegion", region);
+    } else {
+      throw new RuntimeException("IAM authentication requires ElastiCache endpoint");
+    }
+  }
+
+  private void setupTraditionalAuthentication(Properties props) {
+    String username = CACHE_USERNAME.getString(props);
+    String password = CACHE_PASSWORD.getString(props);
+    props.setProperty("cacheUsername", username);
+    props.setProperty("cachePassword", password);
+  }
+
+  private String extractHostnameFromCacheEndpoint(Properties props) {
+    String endpoint = CACHE_RW_ENDPOINT_ADDR.getString(props);
+    if (endpoint != null && endpoint.contains(":")) {
+      return endpoint.split(":")[0];
+    }
+    return null;
+  }
+
+  private boolean isElastiCacheEndpoint(String hostname) {
+    return hostname != null && hostname.endsWith(".cache.amazonaws.com");
+  }
+
+  private boolean hasTraditionalCredentials(Properties props) {
+    String username = CACHE_USERNAME.getString(props);
+    String password = CACHE_PASSWORD.getString(props);
+    return username != null && password != null;
+  }
+
+  private void scheduleReAuthentication() {
+    if (!this.iamAuthEnabled) {
+      return;
+    }
+
+    if (reAuthScheduler == null) {
+      synchronized (this) {
+        if (reAuthScheduler == null) {
+          reAuthScheduler = Executors.newSingleThreadScheduledExecutor(r -> {
```

Comment:
Please make this a singleton thread for the entire JVM process instead of one thread per connection.